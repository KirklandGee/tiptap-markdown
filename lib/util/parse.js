import markdownit from "markdown-it";
import { MarkdownParser, defaultMarkdownParser } from "prosemirror-markdown";


const tokenToNode = {
    blockquote: "blockquote",
    paragraph: "paragraph",
    list_item: "listItem",
    bullet_list: "bulletList",
    ordered_list: "orderedList",
    heading: "heading",
    code_block: "codeBlock",
    fence: "codeBlock",
    hr: "horizontalRule",
    image: "image",
    hardbreak: "hardBreak",
}

const tokenToMark = {
    em: "italic",
    strong: "bold",
    link: "link",
    code_inline: "code",
}

function getTokens(schema) {
    return Object.fromEntries(
        Object.entries(defaultMarkdownParser.tokens)
            .map(([name, definition]) => {
                if(definition.block) {
                    return [name, {
                        ...definition,
                        block: tokenToNode[name]
                    }]
                }
                if(definition.node) {
                    return [name, {
                        ...definition,
                        node: tokenToNode[name]
                    }]
                }
                if(definition.mark) {
                    return [name, {
                        ...definition,
                        mark: tokenToMark[name]
                    }]
                }
            })
            .filter(([name, definition]) => {
                if(definition.block) {
                    return schema.nodes[definition.block];
                }
                if(definition.node) {
                    return schema.nodes[definition.node];
                }
                if(definition.mark) {
                    return schema.marks[definition.mark];
                }
            })
    )
}

export function parse(schema, content) {
    if(typeof content === 'string') {
        const parser = new MarkdownParser(schema, markdownit("commonmark", { html: false }), {
            ...getTokens(schema)
        });
        return parser.parse(content).toJSON();
    }
    return content;
}
