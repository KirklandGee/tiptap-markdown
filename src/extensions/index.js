import Highlight from "./highlight";
import TaskItem from "./task-item";
import TaskList from "./task-list";


export default [
    Highlight,
    TaskItem,
    TaskList,
]
