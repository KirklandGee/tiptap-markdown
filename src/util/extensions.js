

export function createMarkdownExtension(type, options) {
    return {
        type,
        serialize: options.serialize,
        parse: options.parse,
        extend(type, extendOptions) {
            return createMarkdownExtension(type, {
                ...options,
                ...extendOptions,
            });
        },
    }
}

