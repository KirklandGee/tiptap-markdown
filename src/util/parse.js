import markdownit from "markdown-it";
import { MarkdownParser, defaultMarkdownParser } from "prosemirror-markdown";
import { DOMParser } from "prosemirror-model";
import { isVoidElement } from "./dom";

const tokenToNode = {
    blockquote: "blockquote",
    paragraph: "paragraph",
    list_item: "listItem",
    bullet_list: "bulletList",
    ordered_list: "orderedList",
    heading: "heading",
    code_block: "codeBlock",
    fence: "codeBlock",
    hr: "horizontalRule",
    image: "image",
    hardbreak: "hardBreak",
    table: "table",
    tr: "tableRow",
    th: "tableHeader",
    td: "tableCell",
}

const tokenToMark = {
    em: "italic",
    strong: "bold",
    s: "strike",
    link: "link",
    code_inline: "code",
}

function getTokens(schema) {
    const defaultTokens = Object.fromEntries(
        Object.entries(defaultMarkdownParser.tokens)
            .map(([name, definition]) => {
                if(definition.block) {
                    return [name, {
                        ...definition,
                        block: tokenToNode[name]
                    }]
                }
                if(definition.node) {
                    return [name, {
                        ...definition,
                        node: tokenToNode[name]
                    }]
                }
                if(definition.mark) {
                    return [name, {
                        ...definition,
                        mark: tokenToMark[name]
                    }]
                }
            })
    );
    const tokens = {
        ...defaultTokens,
        fence: {
            ...defaultTokens.fence,
            getAttrs: tok => ({ language: tok.info || null }),
        },
        s: { mark:tokenToMark.s },
    };

    return Object.fromEntries(
        Object.entries(tokens).filter(([name, definition]) => {
            if(definition.block) {
                return schema.nodes[definition.block];
            }
            if(definition.node) {
                return schema.nodes[definition.node];
            }
            if(definition.mark) {
                return schema.marks[definition.mark];
            }
        })
    );
}

function elementFromString(value) {
    // add a wrapper to preserve leading and trailing whitespace
    const wrappedValue = `<body>${value}</body>`

    return new window.DOMParser().parseFromString(wrappedValue, 'text/html').body
}

function handleHtmlInline(state, tokens, index) {
    const htmlParser = DOMParser.fromSchema(state.schema);
    const elementStack = [];
    let htmlContent = '';
    let i = index;

    for (; i < tokens.length; i++) {
        let tok = tokens[i];
        htmlContent += tok.content;

        if(tok.type === 'html_inline') {
            const match = tok.content.match(/<([^\/\s>]+)/);
            if(match) {
                const tagName = match[1];
                if(!isVoidElement(tagName)) {
                    elementStack.push(tagName);
                }
            } else {
                elementStack.pop();
            }
        }

        if(!elementStack.length) {
            break;
        }
    }

    const doc = htmlParser.parse(elementFromString(htmlContent));
    const nodes = doc.content.content[0];

    nodes.forEach(node => {
        state.push(node);
    });

    return i;
}


function handleInline(state, token, tokens) {
    const toks = token.children;

    for (let i = 0; i < toks.length; i++) {
        let tok = toks[i]

        if(tok.type === 'html_inline') {
            i = handleHtmlInline(state, toks, i);
            continue;
        }

        const handler = state.tokenHandlers[tok.type]
        if (!handler)
            throw new Error("Token type `" + tok.type + "` not supported by Markdown parser")
        handler(state, tok, toks, i);
    }
}

function parseDOMNode(parser, domNode) {
    const htmlParser = DOMParser.fromSchema(parser.schema);

    const parts = [...domNode.childNodes]
        .map(node => {
            if (node.nodeType === Node.TEXT_NODE) {
                return {
                    type: 'inline',
                    content: node.textContent,
                }
            } else if (node.nodeType === Node.ELEMENT_NODE) {
                const tokens = parser.tokenizer.parse(node.outerHTML, {});
                if (tokens[0].type === 'paragraph_open') {
                    return {
                        type: 'inline',
                        content: node.outerHTML,
                        node,
                    }
                } else if (tokens[0].type === 'html_block') {
                    return {
                        type: 'block',
                        content: node.outerHTML,
                        node,
                    }
                }
            }
        })
        .reduce((res, part) => {
            const lastPart = res[res.length - 1];
            if(lastPart?.type === 'inline' && part.type === 'inline') {
                lastPart.content += part.content;
                return res;
            }
            return [...res, {...part}];
        }, []);

    const parsed = parts
        .reduce((res, part) => {
            const doc = part.type === 'block'
                ? htmlParser.parse(part.node)
                : parser.parse(part.content);
            if(!res) {
                return doc;
            }
            res.content.content.push(...doc.content.content);
            return res;
        }, null);

    return parsed;
}

export function parse(schema, content, { html }) {
    if(typeof content === 'string') {
        const tokenizer = markdownit({ html });
        const parser = new MarkdownParser(schema, tokenizer, {
            ...getTokens(schema)
        });
        parser.tokenHandlers.inline = handleInline;
        parser.tokenHandlers.html_inline = () => {}; // handle in inline
        parser.tokenHandlers.html_block = () => {};

        return parseDOMNode(parser, elementFromString(content)).toJSON();
    }
    return content;
}
