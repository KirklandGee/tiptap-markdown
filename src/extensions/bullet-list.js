import { Node } from "@tiptap/core";
import { createMarkdownExtension } from "../util/extensions";

const BulletList = Node.create({
    name: 'bulletList',
});

export default createMarkdownExtension(BulletList, {
    serialize(state, node)  {
        const marker = null; // todo
        return state.renderList(node, "  ", () => (marker || "*") + " ");
    },
    parse: {
        // handled by markdown-it
    },
});
