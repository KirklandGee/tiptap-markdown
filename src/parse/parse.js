import markdownit from "markdown-it";
import { DOMParser } from "prosemirror-model";
import { elementFromString } from "../util/dom";
import { normalizeDOM } from "./helpers";

export function parse(schema, content, { extensions, ...options }) {
    const {
        html,
        linkify,
        languageClassPrefix,
        inline,
    } = options;

    if(typeof content === 'string') {
        const renderer = markdownit({
            html,
            linkify,
        });

        extensions.forEach(extension => extension.parse?.setup?.call({ schema, options }, renderer));

        const renderedHTML = renderer.render(content);
        const element = elementFromString(renderedHTML);

        extensions.forEach(extension => extension.parse?.updateDOM?.call({ schema, options }, element));

        normalizeDOM(schema, element, { inline });

        return element.innerHTML;
    }
    return content;
}
