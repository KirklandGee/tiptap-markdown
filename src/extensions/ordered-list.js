import { Node } from "@tiptap/core";
import { createMarkdownExtension } from "../util/extensions";

const OrderedList = Node.create({
    name: 'orderedList',
});


export default createMarkdownExtension();
