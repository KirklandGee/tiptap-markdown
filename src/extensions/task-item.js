import { Node } from "@tiptap/core";
import { createMarkdownExtension } from "../util/extensions";

const TaskItem = Node.create({
    name: 'taskItem',
});

export default createMarkdownExtension(TaskItem, {
    serialize(state, node) {
        const check = node.attrs.checked ? '[x]' : '[ ]';
        state.write(`${check} `);
        state.renderContent(node);
    },
    parse: {
        updateDOM(element) {
            [...element.querySelectorAll('.task-list-item')]
                .forEach(item => {
                    const input = item.querySelector('input');
                    item.setAttribute('data-type', 'taskItem');
                    if(input) {
                        item.setAttribute('data-checked', input.checked);
                        input.remove();
                    }
                });
        },
    }
})
