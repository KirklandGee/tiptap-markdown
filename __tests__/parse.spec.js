import extensions from "../src/extensions";
import { parse as baseParse } from '../src/parse/parse';
import { createEditor, nodes, node, inlineNode, dedent } from "./utils";


function parse(content, { html=true, linkify, image, codeBlock, htmlNode } = {}) {
    const editor = createEditor({
        image,
        htmlNode,
        codeBlock,
    });
    return baseParse(editor.schema, content, {
        extensions,
        html,
        linkify,
        languageClassPrefix: codeBlock?.languageClassPrefix,
    });
}


describe('parse', () => {
    describe('marks', () => {
        test('text', () => {
            expect(inlineNode(parse('example'))).toMatchSnapshot();
            expect(inlineNode(parse('http://example.org'))).toMatchSnapshot('link');
        });
        test('bold', () => {
            expect(inlineNode(parse('**example**'))).toMatchSnapshot();
            expect(inlineNode(parse('<b>example</b>'))).toMatchSnapshot('html');
        });
        test('italic', () => {
            expect(inlineNode(parse('*example*'))).toMatchSnapshot();
            expect(inlineNode(parse('<em>example</em>'))).toMatchSnapshot('html');
        });
        test('strike', () => {
            expect(inlineNode(parse('~~example~~'))).toMatchSnapshot();
            expect(inlineNode(parse('<s>example</s>'))).toMatchSnapshot('html');
        });
        test('code', () => {
            expect(inlineNode(parse('`example`'))).toMatchSnapshot();
            expect(inlineNode(parse('<code>example</code>'))).toMatchSnapshot('html');
        });
        test('link', () => {
            expect(inlineNode(parse('[example](http://example.org)'))).toMatchSnapshot();
            expect(inlineNode(parse('<a href="http://example.org">example</a>'))).toMatchSnapshot('html');
        });
        test('link with linkify', () => {
            expect(inlineNode(parse('http://example.org', { linkify:true }))).toMatchSnapshot();
        });
    });
    describe('nodes', () => {
        test('paragraph', () => {
            expect(nodes(parse('example1\n\nexample2'))).toMatchSnapshot();
            expect(nodes(parse('<p>example1</p><p>example2</p>'))).toMatchSnapshot('html');
        });
        test('headings', () => {
            expect(node(parse('# example'))).toMatchSnapshot('h1');
            expect(node(parse('## example'))).toMatchSnapshot('h2');
            expect(node(parse('### example'))).toMatchSnapshot('h3');
            expect(node(parse('#### example'))).toMatchSnapshot('h4');
            expect(node(parse('##### example'))).toMatchSnapshot('h5');
            expect(node(parse('###### example'))).toMatchSnapshot('h6');
            expect(node(parse('<h1>example</h1>'))).toMatchSnapshot('h1 html');
        });
        test('bullet list', () => {
            expect(node(parse('- example1\n\n- example2'))).toMatchSnapshot();
            expect(node(parse('* example1\n\n* example2'))).toMatchSnapshot();
            expect(node(parse('<ul><li>example1</li><li>example2</li></ul>'))).toMatchSnapshot('html');
        });
        test('ordered list', () => {
            expect(node(parse('1. example1\n2. example2'))).toMatchSnapshot();
            expect(node(parse('<ol><li>example1</li><li>example2</li></ol>'))).toMatchSnapshot('html');
        });
        test('fence', () => {
            expect(node(parse('```\nexample\n```'))).toMatchSnapshot();
            expect(node(parse('```js\nexample\n```'))).toMatchSnapshot('lang');
        });
        test('fence with languageClassPrefix', () => {
            expect(node(parse('```js\nexample\n```', { codeBlock: { languageClassPrefix: 'lang-' } }))).toMatchSnapshot();
        })
        test('code block', () => {
            expect(node(parse('    example'))).toMatchSnapshot();
            expect(node(parse('<pre><code>example</code></pre>'))).toMatchSnapshot('html');
        });
        test('image', () => {
            expect(node(parse('![example](example.jpg)'))).toMatchSnapshot();
            expect(node(parse('![example](example.jpg)', { image: { inline: true } }))).toMatchSnapshot('inline');
            expect(node(parse('<img src="example.jpg" alt="example">'))).toMatchSnapshot('html');
        });
        test('hr', () => {
            expect(node(parse('---'))).toMatchSnapshot();
            expect(node(parse('<hr>'))).toMatchSnapshot('html');
        });
        test('hard break', () => {
            expect(node(parse('example1  \nexample2'))).toMatchSnapshot();
            expect(node(parse('example1<br>example2'))).toMatchSnapshot('html');
        });
        test('table', () => {
            expect(node(parse(dedent`
                example1 | example2
                --- | ---
                example3 | example4
            `))).toMatchSnapshot();

            expect(node(parse(dedent`
                <table>
                <thead>
                    <tr>
                        <th>example1</th>
                        <th>example2</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>example3</td>
                        <td>example4</td>
                    </tr>
                </tbody>
                </table>
            `))).toMatchSnapshot('html');
        });
        test('html', () => {
            expect(node(parse('<custom-element>example</custom-element>', {
                htmlNode: {
                    group: 'block',
                    content: 'inline*',
                    parseHTML: () => [{
                        tag: 'custom-element',
                    }],
                },
            }))).toMatchSnapshot();
        });
        test('html inline', () => {
            expect(node(parse('<custom-element></custom-element>', {
                htmlNode: {
                    group: 'inline',
                    inline: true,
                    parseHTML: () => [{
                        tag: 'custom-element',
                    }],
                },
            }))).toMatchSnapshot();
        });
        test('html disabled', () => {
            expect(node(parse('<custom-element></custom-element>', {
                html: false,
                htmlNode: {
                    group: 'block',
                    parseHTML: () => [{
                        tag: 'custom-element',
                    }],
                },
            }))).toMatchSnapshot();
        });
    });
});

