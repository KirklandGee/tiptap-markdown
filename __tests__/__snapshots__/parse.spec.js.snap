// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse marks bold 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "bold",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks bold: html 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "bold",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks code 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "code",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks code: html 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "code",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks italic 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "italic",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks italic: html 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "italic",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks link 1`] = `
Object {
  "marks": Array [
    Object {
      "attrs": Object {
        "href": "http://example.org",
        "target": "_blank",
      },
      "type": "link",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks link: html 1`] = `
Object {
  "marks": Array [
    Object {
      "attrs": Object {
        "href": "http://example.org",
        "target": "_blank",
      },
      "type": "link",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks strike 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "strike",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks strike: html 1`] = `
Object {
  "marks": Array [
    Object {
      "type": "strike",
    },
  ],
  "text": "example",
  "type": "text",
}
`;

exports[`parse marks text 1`] = `
Object {
  "text": "example",
  "type": "text",
}
`;

exports[`parse nodes bullet list 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example1",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example2",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
  ],
  "type": "bulletList",
}
`;

exports[`parse nodes bullet list 2`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example1",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example2",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
  ],
  "type": "bulletList",
}
`;

exports[`parse nodes bullet list: html 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example1",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example2",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
  ],
  "type": "bulletList",
}
`;

exports[`parse nodes code block 1`] = `
Object {
  "attrs": Object {
    "language": null,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "codeBlock",
}
`;

exports[`parse nodes code block: html 1`] = `
Object {
  "attrs": Object {
    "language": null,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "codeBlock",
}
`;

exports[`parse nodes fence 1`] = `
Object {
  "attrs": Object {
    "language": null,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "codeBlock",
}
`;

exports[`parse nodes fence: lang 1`] = `
Object {
  "attrs": Object {
    "language": "js",
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "codeBlock",
}
`;

exports[`parse nodes hard break 1`] = `
Object {
  "content": Array [
    Object {
      "text": "example1",
      "type": "text",
    },
    Object {
      "type": "hardBreak",
    },
    Object {
      "text": "example2",
      "type": "text",
    },
  ],
  "type": "paragraph",
}
`;

exports[`parse nodes hard break: html 1`] = `
Object {
  "content": Array [
    Object {
      "text": "example1",
      "type": "text",
    },
    Object {
      "type": "hardBreak",
    },
    Object {
      "text": "example2",
      "type": "text",
    },
  ],
  "type": "paragraph",
}
`;

exports[`parse nodes headings: h1 1`] = `
Object {
  "attrs": Object {
    "level": 1,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h1 html 1`] = `
Object {
  "attrs": Object {
    "level": 1,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h2 1`] = `
Object {
  "attrs": Object {
    "level": 2,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h3 1`] = `
Object {
  "attrs": Object {
    "level": 3,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h4 1`] = `
Object {
  "attrs": Object {
    "level": 4,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h5 1`] = `
Object {
  "attrs": Object {
    "level": 5,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes headings: h6 1`] = `
Object {
  "attrs": Object {
    "level": 6,
  },
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "heading",
}
`;

exports[`parse nodes hr 1`] = `
Object {
  "type": "horizontalRule",
}
`;

exports[`parse nodes hr: html 1`] = `
Object {
  "type": "horizontalRule",
}
`;

exports[`parse nodes html 1`] = `
Object {
  "content": Array [
    Object {
      "text": "example",
      "type": "text",
    },
  ],
  "type": "html-node",
}
`;

exports[`parse nodes html disabled 1`] = `
Object {
  "content": Array [
    Object {
      "text": "<custom-element></custom-element>",
      "type": "text",
    },
  ],
  "type": "paragraph",
}
`;

exports[`parse nodes html inline 1`] = `
Object {
  "content": Array [
    Object {
      "type": "html-node",
    },
  ],
  "type": "paragraph",
}
`;

exports[`parse nodes image 1`] = `
Object {
  "attrs": Object {
    "alt": "example",
    "src": "example.jpg",
    "title": null,
  },
  "type": "image",
}
`;

exports[`parse nodes image: html 1`] = `
Object {
  "attrs": Object {
    "alt": "example",
    "src": "example.jpg",
    "title": null,
  },
  "type": "image",
}
`;

exports[`parse nodes image: inline 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "alt": "example",
        "src": "example.jpg",
        "title": null,
      },
      "type": "image",
    },
  ],
  "type": "paragraph",
}
`;

exports[`parse nodes ordered list 1`] = `
Object {
  "attrs": Object {
    "start": 1,
  },
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example1",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example2",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
  ],
  "type": "orderedList",
}
`;

exports[`parse nodes ordered list: html 1`] = `
Object {
  "attrs": Object {
    "start": 1,
  },
  "content": Array [
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example1",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
    Object {
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "example2",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "listItem",
    },
  ],
  "type": "orderedList",
}
`;

exports[`parse nodes paragraph 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "text": "example1",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse nodes paragraph: html 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "text": "example1",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse nodes table 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example1",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableHeader",
        },
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example2",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableHeader",
        },
      ],
      "type": "tableRow",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example3",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableCell",
        },
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example4",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableCell",
        },
      ],
      "type": "tableRow",
    },
  ],
  "type": "table",
}
`;

exports[`parse nodes table: html 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example1",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableHeader",
        },
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example2",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableHeader",
        },
      ],
      "type": "tableRow",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example3",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableCell",
        },
        Object {
          "attrs": Object {
            "colspan": 1,
            "colwidth": null,
            "rowspan": 1,
          },
          "content": Array [
            Object {
              "content": Array [
                Object {
                  "text": "example4",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "tableCell",
        },
      ],
      "type": "tableRow",
    },
  ],
  "type": "table",
}
`;
